text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,411 @@
+
+[source]: https://github.com/rust-lang/cargo/issues/11156#issuecomment-1285951209
+
+## Remapping
+
+[rust-lang/cargo#11156](https://github.com/rust-lang/cargo/issues/11156) was originally about remapping the target directory, not about having a central one but reading the issue, there seems to be no needs for more than the simple redefinition of the target directory proposed in this document. In the future, if `CARGO_TARGET_DIR_REMAP` is introduced, it could be used to be the prefix to the target directory like so:
+
+- Set `CARGO_TARGET_DIR_REMAP=/home/user/projects=/tmp/cargo-build`
+- Compile the crate under `/home/user/projects/foo/` **without** `CARGO_TARGET_DIR` set
+- The resulting target directory will be at `/tmp/cargo-build/foo/target`
+
+By making the priority order `CARGO_TARGET_DIR` > `CARGO_TARGET_DIR_REMAP` (when both are absolute paths) we would keep backward compatibility. Or we could disallow having the two set at once, so that they're alternatives and not ordered.
+
+When `CARGO_TARGET_DIR` is relative, the result could be `/tmp/cargo-build/foo/$CARGO_TARGET_DIR`.
+
+Overall, I feel remapping is much harder to implement well and can be added later without interfering with templates in `CARGO_TARGET_DIR` (and without this RFC interfering with remapping), though the design space is probably bigger than the one for this RFC.
+
+## Advantages of using a hash over regular subdirectories
+
+It's possible to achieve most of the same system proposed here by setting a value like `CARGO_TARGET_DIR="/base/dir/{manifest-path-dirs}"`, where a manifest in `/tmp/test1/test2/Cargo.toml` would resolve the build directory to `/base/dir/tmp/test1/test2/`, but most is not *all* of them:
+
+1. Predictable depth of subdirectories: even with a hash, depth can be unpredictable in some way since the untemplated part can be as long as necessary but by using a hash of predictable length, users can prepare for the limits imposed by their filesystem (for example on Windows, some forms of path are quite short): it would be surprising for a project to build when in `/` but not in `/tmp/some/what/deep/inside/folders` for example
+2. More traversals: by using a hash always cut in the same way to form subdirectories, it's easy to plan traversals for the filesystem
+3. Encourage interactions through `cargo-metadata`: by using paths computed through cargo and not easily derivable from the file tree, future tools will be incentivized to work through `cargo-metadata` to find the target directory, widening adoption and making it easier for the cargo team to ensure nothing breaks in subsequent cargo updates
+4. Avoid introducing a strong dependency on only the path: by using a hash, cargo can add element to it to help differentiate builds: for example we could use more parameters in the hash, see the relevant section in Future Possibilites
```

Comment:
These only matter if hashing is being used.  If we give the user a lower level mechanism, then its fine that is predictable and we shouldn't force them to `cargo metadata`.

Reply:
The subsection is about advantages gained from using a hash so I think those deserve to be here ? They're both advantages and disadvantages to a hash, depending on the use case, but they're certainly relevant I think